### PR TITLE
Support configurable CSP reporting percentage

### LINKED ISSUE #14451: Enable CSP violation reporting for a small proportion of visits
## Description

We don't currently enable CSP violation reports on www.mozilla.org. 

This is partly because it's likely to be incredibly noisy with false positives (things like Add Ons will try to inject legitimate things, which get blocked by the rules, and if we get told about every instance of this, that'd be a lot. 

Also, if we choose to use Sentry as our reporting backend (like we do for future.m.o) the volume of false positives across 100% of our traffic would rapidly exhaust the allocated Sentry quota.

However, while exploring django-csp, I found [there is a setting that enables us to report for just X% of violations](https://django-csp.readthedocs.io/en/latest/reports.html).

As a rough ballpark I'm assuming at least two millions pages served per day, so intially setting X to 0.5% would be ~10k pages a day. Worst-case, if every single one of those had a CSP violation on it, that's still well within our quota and would be easy to spot/tune/disable if need be.

---

#### Success Criteria
- [ ] Bedrock has a new setting for the CSP report header, pointing to the default Sentry profile for Dev/Stage/Prod/Test/Demo
- [ ] 0.5% of pages are served with the report-uri header, which points to an appropriate Sentry profile - we can check this by loading ~200 pages with script and checking the headers.

### PR DESCRIPTION
**Pull request by robhudson**
## One-line summary

This changeset adds support for including a CSP report-uri in a certain percentage of pages, so that we can get an idea of whether pages have inappropriate CSP set, but without saturating our logging service, Sentry.

NB: the relevant config to enable this on Dev, Stage and Prod will be in a separate PR, and that work needs to be merged to make this changeset do anything.

## Issue / Bugzilla link

Resolve #14451 

## Testing

I've added a script called `./bin/qa/check_csp_report.py` which we can use to check the proportion of headers being set, either locally or on an actual server.

E.g. with `CSP_REPORT_PERCENTAGE=50`
```
$ ./bin/check_csp_report.py -n 100 http://localhost:8000/en-US/firefox/
Checking http://localhost:8000/en-US/firefox/
100 requests made
defaultdict(<class 'int'>, {'report-uri not present': 51, 'report-uri present': 49})
```

### CODE REVIEW

File: gcp/bedrock-demos/cloudrun/mozorg-demo.env.yaml [hunk 1]
```diff
@@ -11,7 +11,8 @@ ALLOWED_HOSTS: .allizom.org,.moz.works,.run.app
 CONTENT_CARDS_URL: https://www-dev.allizom.org/media/
 CSP_DEFAULT_SRC: "*.allizom.org"
 CSP_EXTRA_FRAME_SRC: "*.mozaws.net,o1069899.sentry.io"
-CSP_REPORT_ENABLE: "True"
+CSP_REPORT_URI: https://o1069899.ingest.us.sentry.io/api/6260338/security/?sentry_key=97ec0cd426714b728e92f3b3aa62f00b
+CSP_REPORT_PERCENTAGE: 75
```
Comment (janbrasna) [inline gcp/bedrock-demos/cloudrun/mozorg-demo.env.yaml:15]: Once there are both enforced and report-only present at the same time, would they share the config (i.e. endpoint and reporting %%), or one or both are gonna need to be set differently? Mostly I'm thinking if it's possible to differentiate the real violations in Sentry from just the "soft" report-only events from the payloads and separate them accordingly, or if there's a need to gather the two in their own sets to differentiate real CSP issues from just future/projected changes' impact.

(Right now there's only the enforced configuration present, do you plan to also include some report-only config with this changeset to factor in the possibility of two sets of reporting prefs?)

BTW what would be the plan for enforced+report-only setup in general? Run report-only just for demo/staging etc. to gather some stats there, or deploy it also to prod with reasonably small %% sample to get real traffic data? (Can the percentage in that case be smaller than 1% or that's the lowest value?)

Comment (robhudson) [reply gcp/bedrock-demos/cloudrun/mozorg-demo.env.yaml:15]: more discussion is needed on these exact questions. I am thinking that keeping everything separated would be cleanest. I'm not sure how Sentry can segregate the different types. I'm also not sure if Sentry can offer 2 separate reporting URLs. But until we have a report-only CSP we can chug along with the enforced one and figure that out later.

File: bedrock/base/middleware.py [hunk 2]
```diff
@@ -320,4 +322,18 @@ def process_response(self, request, response):
...
                     break
 
-        return super().process_response(request, response)
+        response = super().process_response(request, response)
+
+        # Check enforced CSP reporting.
+        report_percentage = getattr(settings, "CONTENT_SECURITY_POLICY", {}).get("REPORT_PERCENTAGE", 100)
+        remove_header = random.randint(0, 99) >= report_percentage
+        if remove_header:
+            del response.headers[csp.HEADER]
```
Comment (janbrasna) [inline bedrock/base/middleware.py:331]: Maybe I'm understanding this completely wrong so please don't take it too seriously as I can be completely off… but this reads to me as "if outside of reporting sample, remove CSP header completely" — which is probably not what is the intent? The reporting sample %% should only filter out the report-to directive from the CSP header, but keep the policy rules still enforced on client.

(While that could make sense for report-only removing the whole header, I still reckon that for debug purposes even the report-only can be sent with no report-to value if that's not against the specs…)

It seems to be confirmed by tests as: https://github.com/mozilla/bedrock/blob/417ab93ebae46acd47d3817585fc46d1f8949989/bedrock/base/tests/test_middleware.py#L336-L340
checks for the whole header not being present in 0% reporting, but the header should be there, just not its `report-to` directive; this also seems to be confirmed by `check_csp_report.py` run in the OP: 
```
100 requests made
defaultdict(<class 'int'>, {'No CSP header': 50, 'report-uri present': 50})
```
compared to the previous PR:

```
100 requests made
defaultdict(<class 'int'>, {'report-uri present': 50, 'report-uri not present': 50})
```

(present vs. not preset, but no **"No CSP header"** which would indicate an issue)

Comment (robhudson) [reply bedrock/base/middleware.py:331]: Excellent catch! Thank you. I had my head in the details too far when I worked on this and you are absolutely correct... we want to clear the `report-uri` value only, not remove the whole header. I'll fix this up soon.

File: gcp/bedrock-demos/cloudrun/mozorg-demo.env.yaml (hunk 1 above)
Comment (janbrasna) [reply gcp/bedrock-demos/cloudrun/mozorg-demo.env.yaml:15]: I tried to search their docs if/how they segregate ingests based on payload [`disposition`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy-Report-Only#disposition) but to no avail. Maybe it's obvious once the data's there, but right now I don't see any good best practice to refer to.

File: bedrock/base/middleware.py (hunk 2 above)
Comment (janbrasna) [reply bedrock/base/middleware.py:331]: 👍 Glad I'm not just seeing things;) BTW I've also used report-to vs. report-url somewhat deliberately, so take that with a pinch of salt, there's probably just the older single header endpoint used, which makes it all a bit easier (not having to wrestle another header or two for the endpoint definition) — i just never remember which is which so I use the directive names sort of interchangeably O;)

Comment (robhudson) [reply bedrock/base/middleware.py]: Updated

Comment (janbrasna) [reply bedrock/base/middleware.py]: One last Q out of ignorance: This is actually already in [`csp/contrib/rate_limiting.py`](https://github.com/mozilla/django-csp/blob/45cb61a5e762d3f4c8ee62e414302e2978445b93/csp/contrib/rate_limiting.py) — wouldn't be better to have that only in one place, either using the contrib rate-limited middleware, or folding it back into the main code somehow? (As the conf value for it, and the docs, present it as a core feature of the package: [`configuration.rst#policy-settings`](https://github.com/mozilla/django-csp/blob/45cb61a5e762d3f4c8ee62e414302e2978445b93/docs/configuration.rst#policy-settings) ~ only when you find your way to [`reports.rst#throttling-the-number-of-reports`](https://github.com/mozilla/django-csp/blob/45cb61a5e762d3f4c8ee62e414302e2978445b93/docs/reports.rst#throttling-the-number-of-reports) that explains the contrib way.)

I understand it was historically outside of the main code for testing in the wild, but this seems as a viable feature long-term, worth keeping in…?

Comment (robhudson) [reply bedrock/base/middleware.py]: That's a good call-out that we should mention the middleware when we mention `REPORT_PERCENTAGE`. I'm hesitant to add more code to the middleware that would run on every request/response if it is unneeded. And it isn't obvious that this is a feature that everyone would need.

I've opened the django-csp issue 232 to address the documentation issue.

As for subclassing the middleware, yes, but there's currently a bug I need to address and roll a new release for. But since we are not reporting yet and we won't be reporting at 100%, I can subclass the middleware to avoid duplication. But this was a good place to test the fix in the meantime. More to come here soon.

Comment (janbrasna) [reply bedrock/base/middleware.py]: I was kinda thinking about resolving the 100<100 issue along the way — I don't know it enough to assess if that makes any perf impact, but my idea was basically (if that was part of the main csp middleware):

- is it set to `0`? skip the rand (rate-limiting) and just always remove the report-to from directives
- is it set to say `>99` or unset/`None`? skip the rand (and rate-limiting code altogether) and break early not touching anything
- only if set to a value that needs the 0..100 rand runs then and only then go ahead and do the rate-limiting dance…

Couldn't that solve both the unnecessary random code if 0/100 clearly set (and fixing the current 100% edge case), and not running the extra code if not configured — thus not needing to keep the feature separately in contrib?

File: bedrock/base/tests/test_middleware.py [hunk 3]
```diff
@@ -10,6 +10,7 @@
 from django.test.utils import override_settings
 from django.urls import reverse
 
+import csp.constants as csp
```
Comment (stevejalim) [inline bedrock/base/tests/test_middleware.py:13]: nit: this renamed import would bite me, I suspect

Comment (robhudson) [reply bedrock/base/tests/test_middleware.py:13]: yeah. Perhaps I should add an import in the root `__init__.py` that imports the constants so an `import csp` will have the constants available. Which might be a nice end-user interface too for the Django settings.

Comment (robhudson) [reply bedrock/base/tests/test_middleware.py]: Updated 👍

File: bedrock/settings/__init__.py [hunk 4]
```diff
@@ -258,15 +259,21 @@
         "connect-src": list(set(_csp_default_src + _csp_connect_src)),
         # support older browsers (mainly Safari)
         "frame-src": _csp_child_src,
-        "report-uri": config("CSP_REPORT_URI", default="") or None,
     },
 }
 
+# Start report-only CSP as a copy. We'll modify it later if needed.
+# Only set up report-only CSP if we have a report-uri set.
+if csp_report_uri := config("CSP_REPORT_URI", default="") or None:
```
Comment (stevejalim) [inline bedrock/settings/__init__.py:267]: Is the `or None` needed here? The default of `""` should be falsey, no?

**Review by stevejalim — APPROVED** [review]
r+wc

File: bedrock/settings/__init__.py (hunk 4 above)
Comment (robhudson) [reply bedrock/settings/__init__.py:267]: django-csp treats all directives that are explicitly `None` as unset. So yeah an empty string will make it not end up in the header, but I'm wanting it more explicit that it's not meant to be set.
